### PR TITLE
Roles Support in QuickPulse Java SDK

### PR DESCRIPTION
**Pull request by rambhatt-msft**
Fix # .

For significant contributions please make sure you have completed the following items:

- [ ] Design discussion issue #
- [ ] Changes in public surface reviewed
- [ ] CHANGELOG.md updated

### CODE REVIEW

File: core/src/main/java/com/microsoft/applicationinsights/internal/quickpulse/DefaultQuickPulseDataFetcher.java
Comment (trask) [inline core/src/main/java/com/microsoft/applicationinsights/internal/quickpulse/DefaultQuickPulseDataFetcher.java:80]: is it same to remove above conditional:

```
if (!LocalStringsUtils.isNullOrEmpty(roleName)) {
    roleName = "\"" + roleName + "\"";
}
```

and do this (more similar to attributes)?

```suggestion
        sb.append("\"RoleName\": \"").append(roleName).append("\",");
```

File: core/src/main/java/com/microsoft/applicationinsights/internal/quickpulse/DefaultQuickPulseDataFetcher.java
Comment (trask) [inline core/src/main/java/com/microsoft/applicationinsights/internal/quickpulse/DefaultQuickPulseDataFetcher.java:103]:
```suggestion
            final String endpoint = LocalStringsUtils.isNullOrEmpty(redirectedEndpoint) ? getQuickPulseEndpoint() : redirectedEndpoint;
            final HttpPost request = networkHelper.buildRequest(currentDate, this.getEndpointUrl(endpoint));
```

File: core/src/main/java/com/microsoft/applicationinsights/internal/quickpulse/DefaultQuickPulseDataFetcher.java
Comment (trask) [inline core/src/main/java/com/microsoft/applicationinsights/internal/quickpulse/DefaultQuickPulseDataFetcher.java:127]: or maybe `endpointPrefix`?
```suggestion
    String getEndpointUrl(String endpoint) {
        return endpoint + "/post?ikey=" + getInstrumentationKey();
```

File: core/src/main/java/com/microsoft/applicationinsights/internal/quickpulse/DefaultQuickPulseDataSender.java
Comment (trask) [inline core/src/main/java/com/microsoft/applicationinsights/internal/quickpulse/DefaultQuickPulseDataSender.java:93]: is this a change in behavior?

File: core/src/main/java/com/microsoft/applicationinsights/internal/quickpulse/DefaultQuickPulsePingSender.java
Comment (trask) [inline core/src/main/java/com/microsoft/applicationinsights/internal/quickpulse/DefaultQuickPulsePingSender.java:78]: same comment as above about removing conditional and changing this, primarily to match the other attributes

```suggestion
                "\"RoleName\":\"" + roleName + "\"," +
```

File: core/src/main/java/com/microsoft/applicationinsights/internal/quickpulse/DefaultQuickPulsePingSender.java
Comment (trask) [inline core/src/main/java/com/microsoft/applicationinsights/internal/quickpulse/DefaultQuickPulsePingSender.java:100]:
```suggestion
        final String endpoint = LocalStringsUtils.isNullOrEmpty(redirectedEndpoint) ? getQuickPulseEndpoint() : redirectedEndpoint;
        final HttpPost request = networkHelper.buildPingRequest(currentDate, getQuickPulsePingUri(endpoint), quickPulseId, machineName, roleName, instanceName);
```

File: core/src/main/java/com/microsoft/applicationinsights/internal/quickpulse/DefaultQuickPulsePingSender.java
Comment (trask) [inline core/src/main/java/com/microsoft/applicationinsights/internal/quickpulse/DefaultQuickPulsePingSender.java:138]: or `endpointPrefix`?
```suggestion
    String getQuickPulsePingUri(String endpoint) {
        return endpoint + "/ping?ikey=" + getInstrumentationKey();
```

File: core/src/main/java/com/microsoft/applicationinsights/internal/quickpulse/QuickPulseHeaderInfo.java
Comment (trask) [inline core/src/main/java/com/microsoft/applicationinsights/internal/quickpulse/QuickPulseHeaderInfo.java:7]:
```suggestion
    private final QuickPulseStatus quickPulseStatus;
    private final String qpsServiceEndpointRedirect;
    private final long qpsServicePollingIntervalMillis;
```

**Review by trask — COMMENTED** [review]
Thanks @rambhatt-msft, this is looking great!

File: core/src/main/java/com/microsoft/applicationinsights/internal/quickpulse/DefaultQuickPulsePingSender.java
Comment (rambhatt-msft) [reply core/src/main/java/com/microsoft/applicationinsights/internal/quickpulse/DefaultQuickPulsePingSender.java:78]: This is a great question 😃 The problem here is that, without this conditional, if the value of `roleName` is `null` then it will become `roleName: "null"` in the resulting json, as in a string called "null" which is incorrect, we want it to be `roleName: null` so the service interprets this correctly.

File: core/src/main/java/com/microsoft/applicationinsights/internal/quickpulse/DefaultQuickPulseDataSender.java
Comment (rambhatt-msft) [reply core/src/main/java/com/microsoft/applicationinsights/internal/quickpulse/DefaultQuickPulseDataSender.java]: Thank you for catching it! This was a mistake on my part

File: core/src/main/java/com/microsoft/applicationinsights/internal/quickpulse/DefaultQuickPulseDataFetcher.java
Comment (rambhatt-msft) [reply core/src/main/java/com/microsoft/applicationinsights/internal/quickpulse/DefaultQuickPulseDataFetcher.java:80]: This is a great question 😃 The problem here is that, without this conditional, if the value of `roleName` is `null` then it will become `roleName: "null"` in the resulting json, as in a string called "null" which is incorrect, we want it to be `roleName: null` so the service interprets this correctly.

File: core/src/main/java/com/microsoft/applicationinsights/internal/quickpulse/QuickPulseNetworkHelper.java
Comment (tokaplan) [inline core/src/main/java/com/microsoft/applicationinsights/internal/quickpulse/QuickPulseNetworkHelper.java:55]: Why 2?

File: core/src/main/java/com/microsoft/applicationinsights/internal/quickpulse/QuickPulseNetworkHelper.java
Comment (tokaplan) [inline core/src/main/java/com/microsoft/applicationinsights/internal/quickpulse/QuickPulseNetworkHelper.java:80]: Are we good on case sensitivity here?

File: core/src/main/java/com/microsoft/applicationinsights/internal/quickpulse/QuickPulseNetworkHelper.java
Comment (rambhatt-msft) [reply core/src/main/java/com/microsoft/applicationinsights/internal/quickpulse/QuickPulseNetworkHelper.java:55]: > Why 2?

checkout DefaultQuickPulsePingSender.java : 76, this is to match whats in the body

`"\"InvariantVersion\": 2," +`

Comment (tokaplan) [reply core/src/main/java/com/microsoft/applicationinsights/internal/quickpulse/QuickPulseNetworkHelper.java:55]: Let's bump it - here and in the body. In fact, can we please have it in a centralized location rather than in 2 separate places? We usually bump the invariant version when we add behaviors.

File: core/src/main/java/com/microsoft/applicationinsights/internal/quickpulse/DefaultQuickPulseDataFetcher.java
Comment (trask) [reply core/src/main/java/com/microsoft/applicationinsights/internal/quickpulse/DefaultQuickPulseDataFetcher.java:80]: oh I see, and the others are never null, makes sense 👍

File: core/src/main/java/com/microsoft/applicationinsights/internal/quickpulse/QuickPulseNetworkHelper.java
Comment (trask) [reply core/src/main/java/com/microsoft/applicationinsights/internal/quickpulse/QuickPulseNetworkHelper.java:80]: 👍 for equalsIgnoreCase

File: core/src/main/java/com/microsoft/applicationinsights/internal/quickpulse/DefaultQuickPulseDataFetcher.java
Comment (rambhatt-msft) [inline core/src/main/java/com/microsoft/applicationinsights/internal/quickpulse/DefaultQuickPulseDataFetcher.java:71]: @trask Do you know why this value and the one below in `DefaultQuickPulsePingSender.java ` are different? They really should not be, as far as I understand.

File: core/src/main/java/com/microsoft/applicationinsights/internal/quickpulse/DefaultQuickPulsePingSender.java
Comment (rambhatt-msft) [inline core/src/main/java/com/microsoft/applicationinsights/internal/quickpulse/DefaultQuickPulsePingSender.java:63]: @trask and here.

File: core/src/main/java/com/microsoft/applicationinsights/internal/quickpulse/DefaultQuickPulseDataFetcher.java
Comment (trask) [reply core/src/main/java/com/microsoft/applicationinsights/internal/quickpulse/DefaultQuickPulseDataFetcher.java:71]: hey @rambhatt-msft, I'm not sure why, it does seem like a mistake that they are different. I looked at the git history and found this: #904. Is it ok for us to go to v2/v3, even though we don't support filtering and sample telemetry?

File: core/src/main/java/com/microsoft/applicationinsights/internal/quickpulse/QuickPulse.java
Comment (trask) [inline core/src/main/java/com/microsoft/applicationinsights/internal/quickpulse/QuickPulse.java:47]: it's more conventional to make this a primitive
```suggestion
    public static final int QP_INVARIANT_VERSION = 1;
```

File: core/src/main/java/com/microsoft/applicationinsights/internal/quickpulse/QuickPulseNetworkHelper.java
Comment (trask) [inline core/src/main/java/com/microsoft/applicationinsights/internal/quickpulse/QuickPulseNetworkHelper.java:53]: if you change above to primitive, then:
```suggestion
        request.addHeader(QPS_INVARIANT_VERSION_HEADER, Integer.toString(QuickPulse.QP_INVARIANT_VERSION));
```

**Review by trask — APPROVED** [review]
👍 will wait for your confirmation before merging

File: core/src/main/java/com/microsoft/applicationinsights/internal/quickpulse/QuickPulseNetworkHelper.java
Comment (rambhatt-msft) [reply core/src/main/java/com/microsoft/applicationinsights/internal/quickpulse/QuickPulseNetworkHelper.java]: thanks for the suggestion

**Comment by rambhatt-msft** [comment]
> 👍 will wait for your confirmation before merging

I am ready with the change. @tokaplan when you feel comfortable with it, we can merge.

File: core/src/main/java/com/microsoft/applicationinsights/internal/quickpulse/DefaultQuickPulseDataFetcher.java
Comment (rambhatt-msft) [reply core/src/main/java/com/microsoft/applicationinsights/internal/quickpulse/DefaultQuickPulseDataFetcher.java:71]: We should keep the version unchanged at 1. Since higher versions, like you said, support filtering and sample telemetry and the version numbers are used to distinguish between them. This change I did now is gonna be in all SDKs so its not a distinguishing feature.

**Review by tokaplan — APPROVED** [review]

**Comment by trask** [comment]
devops please run

**Comment by trask** [comment]
Thanks @rambhatt-msft!

**Comment by rambhatt-msft** [comment]
> Thanks @rambhatt-msft!

Thank for approving @trask  😃

**Pull request merged**
